### PR TITLE
Parser

### PR DESCRIPTION
**Pull request by cabouezzi**
* IExpr interface for expression types following visitor pattern.  
	* Unary, Binary, Literal, Grouping
* ASTPrinter class conforms to visitor interface to pretty print an expression. 
* Created Parser class that currently only parses single expressions
* Added shell script to automatically build and run a test file (named ExpressionTest.lira)

### CODE REVIEW

File: lira/ASTPrinter.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+using System;
+using System.Text;
+
+namespace Lira;
+
+public class ASTPrinter: IExpr.IVisitor<string>
```
Comment (MarioFares) [inline lira/ASTPrinter.cs]: Put a space before the colon.

File: lira/ASTPrinter.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+{
+
+    public ASTPrinter() {}
```
Comment (MarioFares) [inline lira/ASTPrinter.cs]: No need to put this now. An empty public constructor is automatically generated by the compiler if none is provided.

File: lira/ASTPrinter.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        builder.Append(")");
+
+        return builder.ToString();
+    }
+
+    public string VisitBinary(IExpr.Binary binary) => Parenthesize(binary.Operator.Lexeme, binary.Left, binary.Right);
+
+    public string VisitGrouping(IExpr.Grouping grouping) => Parenthesize("group", grouping.Expr);
+
+    public string VisitLiteral(IExpr.Literal literal) => literal.Value == null ? "nil" : $"{literal.Value}";
```
Comment (MarioFares) [inline lira/ASTPrinter.cs]: Use `is null` and `is not null` to check for null.

File: lira/IExpr.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Linq.Expressions;
+
+namespace Lira;
+
+// This is an auto-generated file
```
Comment (MarioFares) [inline lira/IExpr.cs]: Remove.

File: lira/IExpr.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+public interface IExpr
+{	
+	public interface IVisitor<T>
```
Comment (MarioFares) [inline lira/IExpr.cs:5]: Fix the indentation in this one. The indentation in this file is off.

File: lira/Main.cs [hunk 6]
```diff
@@ -28,7 +28,31 @@ public static void RunFile(string path)
...
+        
+        Console.WriteLine("Scanner returned following tokens:");
+        foreach(Token token in tokens) Console.Write($"{token} ");
+
+        Console.WriteLine("\n");
+
+        Parser parser = new(tokens);
+        IExpr? expr = parser.Parse();
+
+        if (expr == null)
```
Comment (MarioFares) [inline lira/Main.cs]: Again, use `is null`.

File: lira/Parser.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
+using System;
+using System.Linq.Expressions;
+
+namespace Lira;
+
+public class Parser
+{
+    public class ParserError: Exception {}
```
Comment (MarioFares) [inline lira/Parser.cs]: Space before colon.

File: .gitignore [hunk 8]
```diff
@@ -6,6 +6,9 @@
 # Chaniel uses mac :)
 .DS_Store
 
+# Project build folder
+.lirabuild
+
```
Comment (MarioFares) [inline .gitignore:11]: You should add `.vscode` to `.gitignore`.

**Review by MarioFares — CHANGES_REQUESTED** [review]

File: lira/IExpr.cs (hunk 4 above)
Comment (cabouezzi) [reply lira/IExpr.cs]: But it is an auto-generated file

**Review by MarioFares — APPROVED** [review]

**Pull request merged**